### ISSUE TITLE
vue-cli-service: command not found

### ISSUE THREAD

**Issue by DimaFromCanada** [vuejs/vue-cli#2404]
### Version
3.0.1

### Node and OS info
nvm 0.33.1 / Node v8.11.4 / Ubuntu 16.04

### Steps to reproduce
npm run build fails because `vue-cli-service: command not found`

### What is expected?
npm run build succeeds

### What is actually happening?
npm run build fails because `vue-cli-service: command not found`

---
Installed globally with `npm install -g @vue/cli@latest`

**Comment by sodatea** [comment]
Please provide a valid reproduction when opening an issue.

**Comment by DimaFromCanada** [comment]
@sodatea please explain how it would be possible to provide a reproduction for this issue.

**Comment by sodatea** [comment]
Like, the project repository created by `vue create`.

**Comment by DimaFromCanada** [comment]
Okay, here is a repository that I created with `vue create`:

https://github.com/DimaFromCanada/testrepoforvue.git

**Comment by DimaFromCanada** [comment]
This works fine on my macOS. The issue is on Ubuntu 16.04 with NVM.

Yes, I have tried `rm -rf node_modules` and reinstall.

**Comment by sodatea** [comment]
Still can't reproduce with a fresh install of `nvm` & `node`.
https://asciinema.org/a/hRDyrM5HcDHVieV3mPLEVlhdL

Maybe the `$PATH` variable is somehow messed up in your environment.

**Comment by DimaFromCanada** [comment]
Please let me know what executables the $PATH should include.

Have you tested in Ubuntu 16.04?

**Comment by sodatea** [comment]
This is most likely an `npm` issue (for not prepending `./node_modules/.bin` to `$PATH`). So please also try updating  `npm` version.

**Comment by sodatea** [comment]
Closing due to inactivity for more than 1 week. Please open a new issue with a reference to this one if you can follow up with more information.

**Comment by patientplatypus** [comment]
Yeah, don't close this issue please. I'm getting this as well. `vue-cli-service: not found`

**Comment by medmin** [comment]
Hi, everyone, I've got the same error.

Apparently, I try to build the app locally, and it works fine.

But when I push the code to github, and the arrival of a new commit triggers travis-ci build job, and it gives me the error which is "/bin/sh: 1: vue-cli-service: not found". 

https://github.com/whizjs/superjs/blob/master/.travis.yml

By the way, travis-ci has been working fine util this latest commit.

**Comment by elhenro** [comment]
same issue here on ubuntu 18.04.. i just tried a complete reinstall of nodejs and npm and deleted all global node_modules

**Comment by ddosdor** [comment]
Hi,

I have exactly the same error for gitlab-ci:

`yarn run v1.12.3
$ vue-cli-service build
/bin/sh: 1: vue-cli-service: not found
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 127.
ERROR: Job failed: exit code 1`

**Comment by GiorgosK** [comment]
Had similar issues as above and this worked 
`sudo npm cache clean -f`

**Comment by diachedelic** [comment]
`rm -rf node_modules package-lock.json && npm install` worked for me

**Comment by sanfx** [comment]
Tried all the options above nothing worked for me

**Comment by PeterA1** [comment]
If you look at https://cli.vuejs.org/guide/cli-service.html it tells you that after installing @vue/cli-service ( which may be part of the standard CLI installation ) then vue-cli-service is available in npm scripts, but that if you want to access it directly then you need to use ./node_modules/.bin/vue-cli-service This worked for me, and I guess you could add ./node_modules/.bin to your $PATH if you were so inclined.

**Comment by medmin** [comment]
I think @PeterA1 could be right.

**Comment by Rozkalns** [comment]
I had an issue when I had NODE_ENV set to production, which was, quite for sure, incorrect for this script to execute.

**Comment by lstanard** [comment]
Tried all of the suggestions above and still having this issue with OS X 10.14.2. `vue --version` displays 3.3.0.

**Comment by ciryon** [comment]
I had the same problem when building on Bitbucket Pipelines. The solution was, similar to above, to go to pipelines, select Caches, mouseover the node_modules dependency cache and click the trashcan icon.

**Comment by barsv** [comment]
on ubuntu I fixed this by:
1) uninstalled my old npm that was installed using apt install
`apt-get remove npm`
2) installed nvm
`curl -o- https://raw.githubusercontent.com/creationix/nvm/v0.33.9/install.sh | bash`
3) restarted terminal
4) installed latest npm using nvm
`nvm install --latest-npm`
5) reinstall packages as suggested by diachedelic
`rm -rf node_modules package-lock.json && npm install`

**Comment by thammarith** [comment]
I got it working with `npx vue-cli-service ...`

**Comment by zacharyyates** [comment]
Not sure if this is the correct solution, but deleting node_modules and npm install didn't solve it for me.  It looks like you need to install @vue/cli-service and any plugins you use globally.  Here's my solution:
Host: Windows 10

Dockerfile
```
FROM node:8.10-stretch

# update npm & install vue cli peer dependencies
RUN npm install -g npm
RUN npm install -g @vue/cli @vue/cli-service @vue/cli-plugin-babel @vue/cli-plugin-eslint

WORKDIR /app

COPY package*.json ./
RUN npm ci

VOLUME /app

CMD npm run serve
```

docker-compose.yml
```
version: "3.5"

services:
  ui:
    build:
      context: ui/
    ports:
      - 3000:80
    volumes:
      - ./ui/:/app
      - shadow-ui:/app/node_modules:delegated # see comment below
    stdin_open: true
    tty: true
    environment:
      CHOKIDAR_USEPOLLING: 'true'

volumes:
  shadow-ui: # see comment below
```

The `shadow-ui` volume prevents the node_modules directory on my host (windows) from being mounted into my container (debian), so npm ci will download the correct binaries for node-sass.  This error presents it's self as "Could not find package vue-template-compiler" or something similar.

**Comment by hyuckgi** [comment]
if happened while building docker image, try to remove `NODE_ENV=production` before `npm install` or `yarn install`.

**Comment by canonigod** [comment]
sudo yarn install fixed mine

**Comment by michealjroberts** [comment]
I think this issue needs to be re-opened because I have got this error simply by deleting my node_modules and re-running npm install from inside the root directory of the application. Running $ vue create project-name allows for npm run serve to work correctly.

**Comment by microJ** [comment]
that fixs me:
edit `package.json` -> `scripts`, change `vue-cli-service` to `node_modules/.bin/vue-cli-service`

**Comment by sivang** [comment]
I also get it, will check if @microJ 's solution fixes it.

**Comment by PeterA1** [comment]
See... 

> Of course the last comment is the one that holds the answer for me. Resolved by prepending my command with `./node_modules/.bin/`

**Comment by Akryum** [comment]
This issue seems to be caused by several user environment related issues:
- corrupted `node_modules`, which can be caused by bad installation of the project packages ➡️ delete `node_module` and reinstall (deleting `package-lock.json` or `yarn.lock` too before reinstalling can also help)
- using yarn to add a dependency in a npm project or using npm to add a dependency in a yarn project (**don't do this**)
- `:` in the path, for example `/Users/julien/programming:work/path/to/my-app` (`PATH` separator is the `:` character)

https://stackoverflow.com/questions/40546231/sh-react-scripts-command-not-found-after-running-npm-start
https://stackoverflow.com/questions/47612580/react-scripts-command-not-found?rq=1
https://github.com/facebook/create-react-app/issues/3256#issuecomment-340259809
https://github.com/facebook/create-react-app/issues/3256#issuecomment-358406611
https://stackoverflow.com/questions/47842849/how-to-resolve-sh-react-scripts-command-not-found

https://npm.community/t/npm-script-cant-find-module-without-the-full-path/1344/15

**Comment by jassehomar** [comment]
> `rm -rf node_modules package-lock.json && npm install` worked for me

this works for me

**Comment by ioslide** [comment]
> `rm -rf node_modules package-lock.json && npm install` 为我工作

thanks

**Comment by Seeker1437** [comment]
I did `npm install` and that was enough for me.

**Comment by events-jonas-chrisw** [comment]
Works if you run the "npm install" in "cmd", but not in powershell.

**Comment by anselmobattisti** [comment]
I'm using ubuntu inside windows 10.
In my case the problem WAS the Malwarebytes :( it was blocking some update in my npm install!

**Comment by jung-hunsoo** [comment]
I got the same error in WSL ubuntu on Win 10.

My solution was to call `./node_modules/.bin/vue-cli-service`; i.e. `./node_modules/.bin/vue-cli-service build --mode development`.

**Comment by 0x77dev** [comment]
Hey!
This is the correct solution: you need to install ` @vue/cli-service ` module.
Using NPM
```console
$ npm i -S @vue/cli-service
```
Using Yarn
```
$ yarn add @vue/cli-service
```

**Comment by 0x77dev** [comment]
> Hey!
> This is the correct solution: you need to install `@vue/cli-service` module.
> Using NPM
> 
> ```
> $ npm i -S @vue/cli-service
> ```
> 
> Using Yarn
> 
> ```
> $ yarn add @vue/cli-service
> ```

Its Worked for me.

**Comment by 0x77dev** [comment]
> > Hey!
> > This is the correct solution: you need to install `@vue/cli-service` module.
> > Using NPM
> > ```
> > $ npm i -S @vue/cli-service
> > ```
> > 
> > 
> > Using Yarn
> > ```
> > $ yarn add @vue/cli-service
> > ```
> 
> Its Worked for me.
If it doesent work, do this again, and call it like this ` ./node_modules/.bin/vue-cli-service `

**Comment by dmonti** [comment]
The only thing that worked for me was going to plugins menu and download '@vue/cli-service' from there

**Comment by svyatoslavratov** [comment]
If you use NPM package manager you need move dependencies in package.json from "devDependencies" to "dependencies". Then NPM will install them and vue-cli-service can build your app.

**Comment by maurelyo** [comment]
npm install -g @vue/cli-service-global

**Comment by 0x77dev** [comment]
> ` npm install -g @vue/cli-service-global `

This is the correct way... Always works for me.

**Comment by nomadcalendar** [comment]
> I had an issue when I had NODE_ENV set to production, which was, quite for sure, incorrect for this script to execute.
> Changed back to development, and everything is fine again.

That resolved it for me

**Comment by HSpens** [comment]
Using `npm` 

Install globally:
`npm install -g @vue/cli-service-global`

Add the directory where npm install globals to PATH and make it persistent in a new terminal by adding it to ~/.bashrc:
`echo $'\nexport PATH="$(npm -g bin)":$PATH' >> ~/.bashrc`

**Comment by Danielg212** [comment]
run it with npx e.g.

npx vue-cli-service build --mode development

**Comment by mklueh** [comment]
in my case I think there was a conflict with yarn and npm.

I did

1. yarn remove global @vue/cli
2. npm cache -f clean 
3. yarn add global @vue/cli

Which worked for me

**Comment by Jaimin-Patel30** [comment]
I also was getting same issue: I resolved it by cleaning cache of npm npm cache clean -f and removing package-lock.json file rm -rf node-modules package-lock.json and start fresh install.

**Comment by kuchaguangjie** [comment]
Got the same issue when deploy vue project in jenkins.

Here is what I did:

Add node's `bin/` dir, to jenkins user's `.bashrc` file.

e.g

```
# node
NODE_HOME=/home/dev/.nvm/versions/node/default
PATH=$NODE_HOME/bin:$PATH
```

**Comment by BonBonSlick** [comment]
The issue still exists. Running everything as Docs say, does not work.

**Comment by Kyocfe** [comment]
Local build is fine.
It is not built on Docker and Travis.
What's wrong?
```dockerfile
# base image
FROM node:11.12.0-alpine as builder
ENV PATH ./node_modules/.bin:$PATH

RUN apk update && apk --no-cache --virtual build-dependencies add python make g++ \
    && npm install \
    && apk del build-dependencies
RUN npm run build
```

- Result
```bash
sh: vue-cli-service: not found
```

- My package.json file
```json
{
  "private": true,
  "scripts": {
    "serve": "./node_modules/.bin/vue-cli-service serve --modern",
    "build": "./node_modules/.bin/vue-cli-service build --modern"
  }
}
```

i can't understand

**Comment by mort3za** [comment]
My problem was in Netlify build.  
By using `YARN_VERSION` set to `1.17.0` in environment variables of Netlify, the problem fixed.

**Comment by robertwt7** [comment]
getting the same one too

**Comment by LucienZhang** [comment]
In my case, the package `@vue/cli-service` is installed in my local node_modules environment, but not my global environment, so it cannot be used as a command. I type `./node_modules/.bin/vue-cli-service serve` and it works.

**Comment by COil** [comment]
In my case, I simply forgot to run `yarn install` before running the build script `yarn build`.

**Comment by tomparle** [comment]
> If you use NPM package manager you need move dependencies in package.json from "devDependencies" to "dependencies". Then NPM on Heroku will install them and vue-cli-service can build your app.

A potentiel issue of this could be that the production build gets heavier because of the additional dependency, which should not be needed in a production, especially if users want to install the app as a PWA.
I could fix this on Heroku by using the heroku build hooks as [explained here](https://devcenter.heroku.com/articles/nodejs-support#customizing-the-build-process)
```
"heroku-prebuild": "npm install -g @vue/cli-service",
```

You can also customize the `heroku-postbuild` if you need additional configuration.

**Comment by johnsampson** [comment]
Similar to @COil's comment.  Simply got this error when cloning new repo and attempting to run `yarn serve` before `yarn`/`yarn install`.

**Comment by sasokovacic** [comment]
My solution on Ubuntu 20.04:

`npm i -g @vue/cli`
`rm -rf node_modules package-lock.json && npm install`

**Comment by songkainpu** [comment]
> ### Version
> 3.0.1
> 
> ### Node and OS info
> nvm 0.33.1 / Node v8.11.4 / Ubuntu 16.04
> 
> ### Steps to reproduce
> npm run build fails because `vue-cli-service: command not found`
> 
> ### What is expected?
> npm run build succeeds
> 
> ### What is actually happening?
> npm run build fails because `vue-cli-service: command not found`
> 
> Installed globally with `npm install -g @vue/cli@latest`
what's the fucking????

**Comment by FErookie** [comment]
> Hey!
> This is the correct solution: you need to install `@vue/cli-service` module.
> Using NPM
> 
> ```
> $ npm i -S @vue/cli-service
> ```
> 
> Using Yarn
> 
> ```
> $ yarn add @vue/cli-service
> ```

it's work for me too

**Comment by philasf** [comment]
I try all of them and now it's working)

**Comment by us3r01** [comment]
> `rm -rf node_modules package-lock.json && npm install` worked for me

This worked for me!

**Comment by TruongDungkyo** [comment]
> `rm -rf node_modules package-lock.json && npm install` worked for me

This worked for me!

**Comment by unikcc** [comment]
> > If you use NPM package manager you need move dependencies in package.json from "devDependencies" to "dependencies". Then NPM on Heroku will install them and vue-cli-service can build your app.
> 
> A potentiel issue of this could be that the production build gets heavier because of the additional dependency, which should not be needed in a production, especially if users want to install the app as a PWA.
> I could fix this on Heroku by using the heroku build hooks as [explained here](https://devcenter.heroku.com/articles/nodejs-support#customizing-the-build-process)
> 
> ```
> "heroku-prebuild": "npm install -g @vue/cli-service",
> ```
> 
> You can also customize the `heroku-postbuild` if you need additional configuration.

First I update the npm to the lasted version, then I run it, This works for me!

**Comment by KrishKayc** [comment]
> `rm -rf node_modules package-lock.json && npm install` worked for me

This worked for me too. Ubuntu 18.04

**Comment by monkut** [comment]
Still an issue for me...

sudo npm install -g @vue/cli
...
$ vue
command "vue" not found

**Comment by youssef-tharwat** [comment]
> I had an issue when I had NODE_ENV set to production, which was, quite for sure, incorrect for this script to execute.
> Changed back to development, and everything is fine again.

Yeah but then how do you deploy your actual production envoirnment then :D ?

**Comment by aleksey-hoffman** [comment]
## Another solution
### If you use `yarn`: 

0. Install the vue cli globally if you didn't already:

```
yarn global add @vue/cli
```

1. Get the path where yarn stores global packages:

```
yarn -g bin

// Output: home/YOUR_USER_NAME/node_modules/.bin
```

2. Open file `~/.bashrc` for editing in any text editor on directly in terminal using `nano`:

```
sudo nano ~/.bashrc
```

3. Add the following line on top. And replace `YOUR_USER_NAME` with your user name.:
```
export PATH=$PATH:/home/YOUR_USER_NAME/.yarn/bin
```
4. Save with `Ctrl + S` and exit `Ctrl + X`

5. Log out / Log in (or just restart the computer)

### If you use `npm`: 

Do the same thing but:
- In the step 0 run `npm install -g @vue/cli` instead of `yarn global add @vue/cli` 
- In the step 1 run `npm -g bin` instead of `yarn -g bin`

--- 

- The other solutions worked for me only temporarily until the terminal window is re-opened.
- I don't use Linux as my primary OS, so I don't know if it's a good idea to add path variables to `~/.bashrc` file.

**Comment by Glen-Bai** [comment]
I got the same issue on macOS.

**Comment by Glen-Bai** [comment]
I got the same issue on macOS.
Here is my solution.
Besides the error message above, there are other tips:
npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/xxx/.npm/_logs/2020-10-20T07_41_00_394Z-debug.log

Open the file of xxx-debug.log:
///
6 verbose cli   '/usr/local/Cellar/node/14.12.0/bin/node',
6 verbose cli   '/Users/xxx/node_modules/.bin/npm', <-------!!! wrong
6 verbose cli   'install',
6 verbose cli   '--loglevel',
6 verbose cli   'error',
6 verbose cli   '@vue/cli-plugin-babel@~4.5.0',
6 verbose cli   '@vue/cli-plugin-eslint@~4.5.0',
6 verbose cli   '@vue/cli-service@~4.5.0',
6 verbose cli   '--save-dev'
6 verbose cli ]
7 info using npm@6.5.0    <-------!!! wrong
8 info using node@v14.12.0
///

I found the npm version and path are wrong.
I open the path of the npm of log file.
cd  /Users/xxx/node_modules/.bin/
then i run the command of ls -al.
///
lrwxr-xr-x    npm -> ../npm/bin/npm-cli.js
lrwxr-xr-x    npx -> ../npm/bin/npx-cli.js
///
The links above point to wrong paths.

The correct paths are :
lrwxr-xr-x    npm -> /usr/local/lib/node_modules/npm/bin/npm-cli.js
lrwxr-xr-x    npx -> /usr/local/lib/node_modules/npm/bin/npm-cli.js

After recorrecting these links. vue works well.

**Comment by kinghat** [comment]
>     * I don't use Linux as my primary OS, so I don't know if it's a good idea to add path variables to `~/.bashrc` file.

for me it has to be in `.bashrc` vs `.bash_profile` for reason i cant remember.

**Comment by redcinelli** [comment]
Hi, I had the same issue for a while.
## TL;DR
```bash
 npm install @vue/cli-service --save-dev
```
As the documentation specify it is a development dependency  https://cli.vuejs.org/guide/#cli-service
> The CLI Service (@vue/cli-service) is a development dependency. It's an npm package installed locally into every project created by @vue/cli.

### Origin
I had a fresh install of nodejs
And just did
```bash
>> sudo npm install -g @vue/cli@latest

>> vue --version
@vue/cli 4.5.8
```
### The issue
The issue presented like this
```bash
>> npm run serve
yarn run v1.22.10
$ vue-cli-service build --mode development --watch
/bin/sh: 1: vue-cli-service: not found
error Command failed with exit code 127.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```
### The fixes:
```bash
npm install @vue/cli-service --save-dev
```
Which led me straight to another error message
```bash
>> npm run serve
yarn run v1.22.10
$ vue-cli-service build --mode development --watch
ERROR  Error: Cannot find module 'vue-template-compiler/package.json'
```
Which I fixed the same way
```bash
npm i vue-template-compiler --save-dev
```

And now it is working fine.

**Comment by dkp1903** [comment]
Those facing this issue making a production build while deploying to Heroku or similar, remove @vue/cli-service and re-add it. Should do the job

**Comment by yoannes** [comment]
Using github actions, the only way that worked for me was to set NODE_ENV to development just for install

```
- run: |
    NODE_ENV=development npm ci
    npm run build
```

**Comment by JTCorrin** [comment]
Thanks @yoannes - I was missing the npm ci as part of my build.

`- run: npm ci && npm run build`

worked for me

**Comment by keyten** [comment]
Getting this issue.

How I got it:
- created empty project via vue-cli-service, added it into backend code and deployed on heroku -- so frontend must be build on the server while deploy. The root folder contains `./client` with the empty project, `scripts.build` in the `./package.json` runs `cd ./client && npm install && npm run build`.
- this error appears

What didn't help:
- deleting package-lock.json
- replacing `vue-cli-service` to either `npx vue-cli-service` or `./node_modules/.bind/vue-cli-service`
```
npm ERR! code E404
npm ERR! 404 Not Found - GET https://registry.npmjs.org/vue-cli-service - Not found
npm ERR! 404 
npm ERR! 404  'vue-cli-service@latest' is not in the npm registry.
npm ERR! 404 You should bug the author to publish it (or use the name yourself!)
```
- adding `@vue/cli-service` to package.json - it's a vue-cli-generated project, it obviously contains `@vue/cli-service` in `devDependencies`
- installing `@vue/cli-service-global` or `vue` globally

**Comment by nguyentruongtansam** [comment]
> This fixed the issue for me:
> 
> ![image](https://user-images.githubusercontent.com/64165327/138573795-a8c34d58-cb1c-4b2e-ad25-99f9b220d6c8.png)
> 
> **Terminal Command:**
> 
> `./node_modules/.bin/vue-cli-service serve`

finally this helped me, thank a lot.

**Comment by enmalafeev** [comment]
@nguyentruongtansam Hi. I wrote like this ./node_modules/.bin/ in npm-sripts.
But on gitlab-ci it alwait fail with error.
> ./node_modules/.bin/vue-cli-service test:unit
sh: 1: ./node_modules/.bin/vue-cli-service: not found
How to fix this on gitlab CI ?

**Comment by kengeene** [comment]
> > `rm -rf node_modules package-lock.json && npm install` worked for me
> 
> This worked for me too. Ubuntu 18.04

This worked for me

**Comment by eggs3** [comment]
> ```shell
>  npm install @vue/cli-service --save-dev
> ```

this worked for me on MacOS 12.3.1

**Comment by xicond** [comment]
In my case,
After tried googling solutions,

https://stackoverflow.com/questions/22343420/npm-not-creating-bin-directory

I found in docker service with NODE_ENV=production it's not compiling the `node_modules/.bin` dir when do `npm i`
so my workaround is to remove that var on the `npm i` become `export NODE_ENV=; npm i; export NODE_ENV=production`

or backup it first to obtain whatever is the previous value
I tried to `echo $NODE_ENV` and it's production

**Comment by eosvn** [comment]
`NAME="Ubuntu"
VERSION="20.04.4 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.4 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal`

Tried all, There's no way it works :(

**Comment by rafaelmils** [comment]
Adding `./node_modules/.bin/` on the command path before the vue-cli-service as you can see in the screenshot below fixed the issue also for me!
<img width="502" alt="Screenshot 2022-10-27 at 11 55 15" src="https://user-images.githubusercontent.com/64504654/198253984-9aca4162-68bc-4f09-8fab-c8eec99ff890.png">
